tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_PrimitiveFromConservative.cpp

```diff
@@ -0,0 +1,282 @@
+      dot_product(spatial_velocity_oneform, expected_spatial_velocity);
+
+  auto tilde_d = make_with_value<Scalar<DataType>>(used_for_size, 0.);
+  auto tilde_tau = make_with_value<Scalar<DataType>>(used_for_size, 0.);
+  auto tilde_s = make_with_value<tnsr::i<DataType, Dim>>(used_for_size, 0.);
+
+  RelativisticEuler::Valencia::conservative_from_primitive(
+      make_not_null(&tilde_d), make_not_null(&tilde_tau),
+      make_not_null(&tilde_s), expected_rest_mass_density,
+      expected_specific_internal_energy, spatial_velocity_oneform,
+      spatial_velocity_squared, expected_lorentz_factor,
+      expected_specific_enthalpy, expected_pressure, sqrt_det_spatial_metric);
+
+  auto rest_mass_density = make_with_value<Scalar<DataType>>(used_for_size, 0.);
+  auto specific_internal_energy =
+      make_with_value<Scalar<DataType>>(used_for_size, 0.);
+  auto lorentz_factor = make_with_value<Scalar<DataType>>(used_for_size, 0.);
+  auto specific_enthalpy = make_with_value<Scalar<DataType>>(used_for_size, 0.);
+  auto pressure = make_with_value<Scalar<DataType>>(used_for_size, 0.);
+  auto spatial_velocity =
+      make_with_value<tnsr::I<DataType, Dim>>(used_for_size, 0.);
+
+  RelativisticEuler::Valencia::primitive_from_conservative<ThermodynamicDim>(
+      make_not_null(&rest_mass_density),
+      make_not_null(&specific_internal_energy), make_not_null(&lorentz_factor),
+      make_not_null(&specific_enthalpy), make_not_null(&pressure),
+      make_not_null(&spatial_velocity), tilde_d, tilde_tau, tilde_s,
+      inv_spatial_metric, sqrt_det_spatial_metric, equation_of_state);
+
+  Approx larger_approx =
+      Approx::custom().epsilon(std::numeric_limits<double>::epsilon() * 1.e6);
+  CHECK_ITERABLE_CUSTOM_APPROX(expected_rest_mass_density, rest_mass_density,
+                               larger_approx);
+  CHECK_ITERABLE_CUSTOM_APPROX(expected_specific_internal_energy,
+                               specific_internal_energy, larger_approx);
+  CHECK_ITERABLE_CUSTOM_APPROX(expected_lorentz_factor, lorentz_factor,
+                               larger_approx);
+  CHECK_ITERABLE_CUSTOM_APPROX(expected_specific_enthalpy, specific_enthalpy,
+                               larger_approx);
+  CHECK_ITERABLE_CUSTOM_APPROX(expected_pressure, pressure, larger_approx);
+  CHECK_ITERABLE_CUSTOM_APPROX(expected_spatial_velocity, spatial_velocity,
+                               larger_approx);
+}
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.RelativisticEuler.Valencia.PrimitiveFromConservative",
+                  "[Unit][RelativisticEuler]") {
+  std::random_device r;
+  const auto seed = r();
+  std::mt19937 generator(seed);
+  INFO("seed = " << seed);
+
+  EquationsOfState::PolytropicFluid<true> polytropic_fluid(100.0, 2.0);
+  EquationsOfState::IdealFluid<true> ideal_fluid(4.0 / 3.0);
+  const double d = std::numeric_limits<double>::signaling_NaN();
+  test_primitive_from_conservative<1, 1>(&generator, polytropic_fluid, d);
```

Comment:
The second template parameter should be inferred, no?

Reply:
@nilsdeppe:
       `candidate template ignored: could not match 'EquationOfState' against 'PolytropicFluid'`

Reply:
Oh well...